client/datasync-client/src/DataManager.ts

```diff
@@ -113,17 +132,73 @@ export class DataManager {
    *
    * This will stop the sync Client API from sending/recieving updates from the remote server.
    *
-   * @returns {*}
+   * @returns Promise (bluebird)
+   */
+  public stop(): Bluebird<any> {
+    const self = this;
+    return new Bluebird(function(resolve, reject) {
+      syncApi.stopSync(self.datasetId, function() {
+        resolve();
+      }, function(error) {
+        reject(new Error(error));
+      });
+    });
+  }
+
+  /**
+   * Forcing the sync framework to do a sync request to the remote server to exchange data.
+   *
+   * @returns Promise (bluebird)
    */
-  public stop(callback: (err?: Error, result?: any) => void) {
-    syncApi.startSync(this.datasetId, function() {
-      return callback();
-    }, function(error) {
-      return callback(error);
+  public forceSync(): Bluebird<any> {
+    const self = this;
+    return new Bluebird(function(resolve, reject) {
+      syncApi.forceSync(self.datasetId, function() {
+        resolve();
+      }, function(error) {
+        reject(new Error(error));
+      });
     });
   }
 
   /**
+   * A utility function to push any updates to the remote server and then stop syncing.
+   * If there are pending sync operations, then force them to sync and then stop syncing.
+   * @param userOptions - object that can contain delay
+   * @returns {*}
+   */
+  public safeStop = function(userOptions) {
+    const self = this;
+    const defaultOptions = {
+      delay: 5000
+    };
+
+    const options = _.defaults(userOptions, defaultOptions);
+
+    function forceSyncThenStop(pendingUpdateQueueSize) {
+      if (pendingUpdateQueueSize === 0) {
+        self.stop().then(Bluebird.resolve);
+        return;
+      }
+      return self.forceSync()
+        .delay(options.delay)
+        .then(self.getQueueSize.bind(self))
+        .then(function(size) {
+          if (size > 0) {
+            Bluebird.reject(new Error('forceSync failed, outstanding results still present'));
+          }
+        })
+        .then(self.stop.bind(self))
+        .then(function() {
+          Bluebird.resolve();
+        }, function() {
+          Bluebird.reject(new Error('forceSync error'));
+        });
```

Comment:
I think this last `then` is redundant, and wrong since it's not `return`ing anything.

If you're just looking to mask the return value and error value, `.then(() => undefined).catch(() => Bluebird.reject(new Error('forceSync error')))` should do the trick.